### ISSUE TITLE
Allow `--flag:any = <default>` to avoid early type restriction

### ISSUE THREAD

**Issue by bobhy** [nushell/nushell#9090]
### Describe the bug

Right now, if you declare a flag or positional of type `Any` but provide a default of a specific type, the parser complains if you try to invoke the custom command with any other type.

### How to reproduce

```
〉cat touch.nu
export def main [ --date (-d):any = 0day ] {
 let touch_date = match ($date | describe) {
     'datetime' => { $date },
     'duration' => {(date now) - $date},
     _          => { 
                    let span = (metadata $date).span;
                    error make {msg:"Incorrect type"
                                label: {text:"Must be datetime or duration"
                                        start: $span.start
                                        end: $span.end
                                        }
                                }
                },
    }
 $touch_date
}

〉use touch.nu

〉touch
Wed, 03 May 2023 10:04:42 -0400 (now)
〉touch --date 1day
Tue, 02 May 2023 10:04:51 -0400 (a day ago)
〉touch --date 2023-05-01
Error: nu::parser::parse_mismatch

  × Parse mismatch during operation.
   ╭─[entry #81:1:1]
 1 │ touch --date 2023-05-01
   ·              ─────┬────
   ·                   ╰── expected duration with valid units
   ╰────

〉touch --date "gronk"
Error: nu::parser::parse_mismatch

  × Parse mismatch during operation.
   ╭─[entry #82:1:1]
 1 │ touch --date "gronk"
   ·              ───┬───
   ·                 ╰── expected duration with valid units
   ╰────
```

Parser has set the type of the flag as `Duration` in the signature, 
```
〉touch --help
Usage:
  > main {flags} 

Flags:
  -d, --date <Duration> -  (default: coming soon)
  -h, --help - Display the help message for this command

```

### Expected behavior

I was expecting the type of flag or positional to be what I specified in the custom command declaration.  
If I don't declare a type but do provide a default, I'd expect the type of the param to become that of the default.
If I declare both type of the parameter and provide a default, I'd expect a parse error about the type mismatch.

### Configuration

| key                | value                                    |
| ------------------ | ---------------------------------------- |
| version            | 0.79.1                                   |
| branch             | help_help                                |
| commit_hash        | 0e07c2c70909df961fc4fa30238d2218a4b14238 |
| build_os           | linux-x86_64                             |
| build_target       | x86_64-unknown-linux-gnu                 |
| rust_version       | rustc 1.67.1 (d5a82bbd2 2023-02-07)      |
| rust_channel       | 1.67.1-x86_64-unknown-linux-gnu          |
| cargo_version      | cargo 1.67.1 (8ecd4f20a 2023-01-10)      |
| build_time         | 2023-04-27 18:26:37 -04:00               |
| build_rust_channel | debug                                    |
| features           | default, zip                             |
| installed_plugins  |                                          |

### Additional context

Related to discussion in https://github.com/nushell/nushell/pull/8858#issuecomment-1532018742

**Comment by fdncred** [comment]
Just hacking around here.
```
> def test [--date = 0day] { echo $"((date now) - $date)" }
> test
Wed, 03 May 2023 10:01:29 -0500 (now)
> test --date 1day
Tue, 02 May 2023 10:01:52 -0500 (a day ago)
> test --date ('2023-05-01' | into datetime)
2day 10hr 2min 9sec 352ms 468µs 200ns
```

**Comment by fdncred** [comment]
It's a little bit confusing that we have
```rust
b"date" => Type::Date
```
and
```rust
Type::Date => SyntaxShape::DateTime
```
and also
```rust
FlatShape::DateTime
Expr::DateTime
SyntaxShape::DateTime //mentioned above
```
but this as well
```rust
Value::Date
Type::Date
Category::DateTime
```

Seems like we should make up our mind whether we're calling things Date or DateTime. I *think* all these above are date time representations.

**Comment by amtoine** [comment]
@fdncred 
agree this is strange, at least confusing :thinking: 

`DateTime` looks good to me :+1:

**Comment by bobhy** [comment]
@fncred those are legit concerns and worth consolidating, but my concern on this issue is the early type binding for parameters declared type `any`.  I can repro that issue with other types:

```
〉def foo [arg:number = 10] { echo $arg }
〉foo 10.3
Error: nu::parser::parse_mismatch

  × Parse mismatch during operation.
   ╭─[entry #37:1:1]
 1 │ foo 10.3
   ·     ──┬─
   ·       ╰── expected int
   ╰────

```
This could be fixed in general by making `any` a union type, though Nu doesn't support that anywhere else.  However, the problem also repros with `number` (as above), which is already a sort of union type.

**Comment by fdncred** [comment]
@bobhy yup, sorry to hijack your issue. There is some funkiness that you've pointed out that needs to be fixed.

**Comment by kubouch** [comment]
+1 for consolidating. While `date` is shorter, I'd probably vote for `datetime` since it's more descriptive. 

As for the original issue, the solution could be along the lines of calling `eval_constant()` on the default argument during parsing and ifer the type from the result (on success, on failure, it would be "any" type), then update the type in the call's signature. `eval_constant()` could also be extended to allow evaluating more stuff at parse time.

**Comment by bobhy** [comment]
Sounds like the above would help in the case that user doesn't declare a type but does provide a default.  
I don't see that it would give an error if user provided (conflicting) info for both; that really feels like it deserves a parse error to me.

**Comment by WindSoilder** [comment]
Oh...just found that it's the same to this issue: #10410
And it's resolved by #10424
